File: E-mart_java/Backend/src/main/java/com/example/Services/ProductService.java
package com.example.Services;

import com.example.Models.Product;
import java.util.List;
import java.util.Optional;

public interface ProductService {

    List<Product> getAllProducts();

    Optional<Product> getProductById(int productid);

    List<Product> getProductsByCategory(int categoryid);

    List<Product> getProductsBySubcategory(int subcategoryid);

//    List<Product> getProductsByBrand(String brand);
//
//    List<Product> getDeals();


}

File: E-mart_.Net/Backend/Services/ProductService.cs
using E_mart.Models;
using E_mart.Repositories;
using Microsoft.EntityFrameworkCore;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace E_mart.Services
{
    public class ProductService : IProductService
    {
        private readonly eMartDbContext _context;

        public ProductService(eMartDbContext context)
        {
            _context = context;
        }

        public async Task<IEnumerable<Product>> GetAllProducts()
        {
            return await _context.Products.ToListAsync();
        }

        public async Task<Product> GetProductById(int id)
        {
            return await _context.Products.FindAsync(id);
            //return await _context.Products.FirstOrDefaultAsync(p => p.Productid == id);
        }

        public async Task<IEnumerable<Product>> GetProductsByCategory(int categoryid)
        {
            return await _context.Products
                                 .Where(p => p.Categoryid == categoryid)
                                 .ToListAsync();
        }

        public async Task<IEnumerable<Product>> GetProductsBySubcategory(int subcategoryid)
        {
            return await _context.Products
                                 .Where(p => p.Subcategoryid == subcategoryid)
                                 .ToListAsync();
        }
    }
}
